L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc

```diff
@@ -0,0 +1,193 @@
+  }
+
+  if (modSum > 0) {
+    //normalize inputs to module sum
+    for (int i = 0; i < 48; i++) {
+      ae_inputArray[i] = mipPt_[i] / modSum;
+
+      //round to precision of input, if bitsPerInput_ is -1 keep full precision
+      if (bitsPerInput_ > 0) {
+        ae_inputArray[i] = round(ae_inputArray[i] * pow(2., bitsPerInput_)) / pow(2., bitsPerInput_);
+      }
+    }
+  }
+
+  tensorflow::Tensor encoder_input(tensorflow::DT_FLOAT, {1, 4, 4, 3});
+  float* d = encoder_input.flat<float>().data();
+  for (int i = 0; i < 48; i++, d++) {
+    *d = ae_inputArray[i];
+  }
+
+  int graphIndex = linkToGraphMap_.at(nLinks);
+
+  std::vector<tensorflow::Tensor> encoder_outputs;
+  tensorflow::run(session_encoder_.at(graphIndex),
+                  {{inputTensorName_encoder_, encoder_input}},
+                  {outputTensorName_encoder_},
+                  &encoder_outputs);
+
+  double ae_encodedLayer[16];
+  d = encoder_outputs[0].flat<float>().data();
+  for (int i = 0; i < encoder_outputs[0].NumElements(); i++, d++) {
+    ae_encodedLayer[i] = *d;
+    //truncate the encoded layer bits
+    if (bitsPerOutput > 0) {
+      ae_encodedLayer[i] = round(ae_encodedLayer[i] * pow(2, bitsPerOutput)) / pow(2., bitsPerOutput);
+    }
+  }
+
+  tensorflow::Tensor decoder_input(tensorflow::DT_FLOAT, {1, 16});
+  d = decoder_input.flat<float>().data();
+  for (int i = 0; i < 16; i++, d++) {
+    *d = ae_encodedLayer[i];
+  }
+
+  std::vector<tensorflow::Tensor> decoder_outputs;
+  tensorflow::run(session_decoder_.at(graphIndex),
+                  {{inputTensorName_decoder_, decoder_input}},
+                  {outputTensorName_decoder_},
+                  &decoder_outputs);
+
+  double ae_outputArray[48];
+
+  d = decoder_outputs[0].flat<float>().data();
+  for (int i = 0; i < decoder_outputs[0].NumElements(); i++, d++) {
+    ae_outputArray[i] = *d;
+  }
+
+  // Add data back into trigger cells
+  if (modSum > 0) {
+    //get detID for everything but cell, take first entry detID and subtract off cellU and cellV contribution
+    HGCalTriggerDetId id(trigCellVecInput.at(0).detId());
+    int cellU_ = id.triggerCellU();
+    int cellV_ = id.triggerCellV();
+    unsigned _id_waferBase = trigCellVecInput.at(0).detId() - (cellU_ << HGCalTriggerDetId::kHGCalCellUOffset) -
+                             (cellV_ << HGCalTriggerDetId::kHGCalCellVOffset);
+
+    //use first TC to find mipPt conversions to Et and ADC
+    float mipPtToEt_conv = trigCellVecInput.at(0).et() / trigCellVecInput.at(0).mipPt();
+    float mipPtToADC_conv = trigCellVecInput.at(0).hwPt() / trigCellVecInput.at(0).mipPt();
+
+    for (int i = 0; i < 48; i++) {
+      if (ae_outputArray[i] > 0) {
+        cellU_ = ae_outputCellU_[i];
+        cellV_ = ae_outputCellV_[i];
+
+        //find detID for this cell
+        unsigned detID = _id_waferBase + (cellU_ << HGCalTriggerDetId::kHGCalCellUOffset) +
+                         (cellV_ << HGCalTriggerDetId::kHGCalCellVOffset);
```

Comment:
Instead of these detid manipulations, it would be clearer to directly create a `HGCalTriggerDetid` from `int subdet, int zp, int type, int layer, int waferU, int waferV, int cellU, int cellV`